### ISSUE TITLE
[AI] Active Learning Pipeline for Continuous Classifier Improvement

### ISSUE THREAD

**Issue by Daksh7785** [riteshbonthalakoti/HELPDESK.AI#1931]
<html>
<body>
<html><head></head><body><h1>🚀 [AI] Active Learning Pipeline for Continuous Classifier Improvement</h1><h2>📌 Overview</h2><p>HelpDesk.AI currently uses a DistilBERT-based ticket classification system that achieves approximately <strong>85% accuracy</strong> during evaluation. While this performance is acceptable initially, the classifier gradually degrades in production due to evolving customer behavior, changing ticket distributions, new terminology, and domain-specific language.</p><p>Although administrator corrections are already being recorded in <code inline="">corrections_log.json</code>, these valuable feedback signals are never used to improve the model. As a result, the system continuously accumulates labeled training data without any mechanism to learn from it.</p><p>This issue proposes implementing a complete <strong>Active Learning &amp; Continuous Retraining Pipeline</strong> that automatically leverages real-world corrections, monitors model performance, evaluates candidate models, and safely deploys improvements.</p><hr><h1>🎯 Problem Statement</h1><p>The current classifier architecture suffers from several machine learning lifecycle limitations.</p><h2>Current Challenges</h2><h3>1. Data Drift</h3><p>Production ticket distributions evolve over time.</p><p>Examples:</p><ul><li><p>New product releases introduce new issue categories.</p></li><li><p>Customer terminology changes.</p></li><li><p>Organizations use domain-specific language not present during initial training.</p></li></ul><p>As a result, prediction quality deteriorates.</p><hr><h3>2. No Continuous Learning</h3><p>The model remains static after deployment.</p><p>Current workflow:</p><pre><code class="language-text">Train Once → Deploy → Never Update
</code></pre><p>Desired workflow:</p><pre><code class="language-text">Train → Deploy → Collect Feedback → Retrain → Improve → Redeploy
</code></pre><p>Without retraining, the classifier cannot adapt to new information.</p><hr><h3>3. Correction Data Is Ignored</h3><p>Administrators frequently correct misclassified tickets.</p><p>Example:</p><pre><code class="language-json">{
  "ticket": "VPN access not working",
  "predicted": "Software Issue",
  "corrected": "Network Issue"
}
</code></pre><p>These corrections represent verified ground-truth labels but are currently unused.</p><hr><h3>4. Accuracy Degradation</h3><p>As data drift increases:</p><ul><li><p>Misclassification rates increase</p></li><li><p>Ticket routing quality decreases</p></li><li><p>Manual intervention increases</p></li><li><p>User trust decreases</p></li></ul><hr><h3>5. No Model Governance</h3><p>The platform lacks:</p><ul><li><p>Model versioning</p></li><li><p>Performance tracking</p></li><li><p>Rollback mechanisms</p></li><li><p>Controlled deployments</p></li></ul><p>This creates operational risk when updating ML models.</p><hr><h1>🌍 Why This Matters</h1><h2>Business Impact</h2><p>Poor classification directly affects:</p><ul><li><p>Ticket routing accuracy</p></li><li><p>SLA performance</p></li><li><p>Support efficiency</p></li><li><p>Customer satisfaction</p></li></ul><p>Even a small drop in accuracy can significantly increase operational overhead.</p><hr><h2>Competitive Advantage</h2><p>Modern AI-powered helpdesk platforms continuously improve through feedback loops.</p><p>Industry leaders use:</p><ul><li><p>Active learning</p></li><li><p>Human-in-the-loop feedback</p></li><li><p>Continuous retraining</p></li><li><p>Performance monitoring</p></li></ul><p>HelpDesk.AI should provide similar capabilities.</p><hr><h2>Operational Efficiency</h2><p>Benefits include:</p><ul><li><p>Fewer manual corrections</p></li><li><p>Better ticket categorization</p></li><li><p>Faster resolution times</p></li><li><p>Reduced support costs</p></li></ul><hr><h1>💥 Real-World Impact</h1><h2>Scenario 1 — Product Terminology Changes</h2><p>A company launches a new feature.</p><p>Users begin submitting tickets containing new terminology.</p><h3>Current Behavior</h3><p>Classifier has never seen these terms.</p><p>Result:</p><ul><li><p>Low confidence predictions</p></li><li><p>Incorrect categories</p></li><li><p>Increased admin corrections</p></li></ul><h3>Desired Behavior</h3><p>Corrections are automatically incorporated into future training cycles.</p><p>The classifier learns new terminology over time.</p><hr><h2>Scenario 2 — Domain-Specific Customers</h2><p>Healthcare customers submit tickets such as:</p><pre><code class="language-text">EMR synchronization failure
</code></pre><p>Financial customers submit:</p><pre><code class="language-text">SWIFT transfer processing error
</code></pre><p>The generic model may struggle with industry-specific language.</p><p>Continuous learning improves classification quality.</p><hr><h2>Scenario 3 — Repeated Misclassifications</h2><p>Administrators repeatedly correct the same prediction errors.</p><p>Current state:</p><pre><code class="language-text">Wrong Prediction
→ Admin Corrects
→ Model Learns Nothing
</code></pre><p>Desired state:</p><pre><code class="language-text">Wrong Prediction
→ Admin Corrects
→ Added To Training Set
→ Future Improvement
</code></pre><hr><h1>✅ Proposed Solution</h1><p>Implement a complete Active Learning and Continuous Retraining Framework.</p><hr><h1>Phase 1: Feedback Collection</h1><h2>Enhanced Correction Logging</h2><p>Extend existing correction logging with additional metadata.</p><h3>Current Data</h3><pre><code class="language-json">{
  "ticket_text": "...",
  "predicted_label": "...",
  "corrected_label": "..."
}
</code></pre><h3>Enhanced Data</h3><pre><code class="language-json">{
  "ticket_text": "...",
  "predicted_label": "...",
  "corrected_label": "...",
  "confidence_score": 0.62,
  "classifier_version": "v1.4.0",
  "tenant_id": "xyz",
  "timestamp": "..."
}
</code></pre><hr><h2>Additional Telemetry</h2><p>Track:</p><ul><li><p>Prediction confidence</p></li><li><p>Model version</p></li><li><p>Correction frequency</p></li><li><p>Category drift</p></li><li><p>User feedback ratings</p></li></ul><p>This data enables smarter retraining decisions.</p><hr><h1>Phase 2: Training Dataset Preparation</h1><p>Create automated preprocessing pipeline.</p><h2>Weekly Data Extraction</h2><p>Scheduled job:</p><pre><code class="language-text">Every Sunday
</code></pre><p>Tasks:</p><ul><li><p>Load recent corrections</p></li><li><p>Validate labels</p></li><li><p>Remove duplicates</p></li><li><p>Remove noisy samples</p></li></ul><hr><h2>Hard Negative Mining</h2><p>Identify:</p><pre><code class="language-text">High Confidence
+
Wrong Prediction
</code></pre><p>Example:</p><pre><code class="language-text">Confidence: 98%
Prediction: Billing
Correct Label: Technical
</code></pre><p>These examples provide strong learning signals.</p><hr><h2>Dataset Balancing</h2><p>Prevent overrepresentation of popular classes.</p><p>Techniques:</p><ul><li><p>Oversampling minority classes</p></li><li><p>Undersampling dominant classes</p></li><li><p>Weighted loss functions</p></li></ul><hr><h2>Annotation Conflict Resolution</h2><p>Handle cases where:</p><pre><code class="language-text">Admin A → Network
Admin B → Software
</code></pre><p>Resolution strategies:</p><ul><li><p>Majority voting</p></li><li><p>Human review</p></li><li><p>Confidence weighting</p></li></ul><hr><h1>Phase 3: Automated Retraining Pipeline</h1><h2>Scheduled Retraining</h2><p>Frequency:</p><pre><code class="language-text">Bi-Weekly
</code></pre><p>GitHub Action:</p><pre><code class="language-bash">.github/workflows/retrain-classifier.yml
</code></pre><hr><h2>Training Process</h2><p>Steps:</p><ol><li><p>Load baseline model</p></li><li><p>Merge correction dataset</p></li><li><p>Fine-tune DistilBERT</p></li><li><p>Generate evaluation metrics</p></li><li><p>Compare against production model</p></li></ol><hr><h2>Evaluation Metrics</h2><p>Track:</p><h3>Classification Metrics</h3><ul><li><p>Accuracy</p></li><li><p>Precision</p></li><li><p>Recall</p></li><li><p>F1 Score</p></li></ul><h3>Operational Metrics</h3><ul><li><p>Misclassification Rate</p></li><li><p>Routing Accuracy</p></li><li><p>Confidence Calibration</p></li></ul><hr><h1>Phase 4: Model Validation Gate</h1><p>Prevent deployment of worse models.</p><h2>Promotion Rules</h2><p>Deploy only if:</p><pre><code class="language-text">Accuracy Gain ≥ 2%
AND
No Major Regression
</code></pre><p>Example:</p>
Model | Accuracy
-- | --
Current | 85%
Candidate | 88%

<hr><h1>🔒 Risk Mitigation</h1><h3>Prevent Model Regression</h3><ul><li><p>Validation thresholds</p></li><li><p>Shadow evaluation</p></li><li><p>Automated testing</p></li></ul><h3>Prevent Data Poisoning</h3><ul><li><p>Correction validation</p></li><li><p>Noise filtering</p></li><li><p>Outlier detection</p></li></ul><h3>Ensure Rollback Safety</h3><ul><li><p>Version snapshots</p></li><li><p>Deployment history</p></li><li><p>One-click rollback</p></li></ul><hr><h1>✅ Acceptance Criteria</h1><h2>Data Collection</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Correction data collected with confidence scores</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Prediction telemetry logged</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model version metadata stored</p></li></ul><h2>Data Preparation</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Weekly correction extraction job implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Hard negatives identified automatically</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Dataset balancing implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Noisy annotations filtered</p></li></ul><h2>Retraining</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Bi-weekly retraining job runs automatically</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> DistilBERT fine-tuning pipeline implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Validation metrics generated after training</p></li></ul><h2>Evaluation</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Candidate model compared against production</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Shadow evaluation runs on live traffic</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Deployment blocked if improvement &lt; 2%</p></li></ul><h2>Model Governance</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Version history maintained</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Rollback capability implemented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Training metadata stored</p></li></ul><h2>Active Learning</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Low-confidence predictions identified</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Feedback requests prioritized intelligently</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Hard negatives incorporated into training</p></li></ul><h2>Enterprise Features</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Tenant-specific model support available</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Fallback to global model implemented</p></li></ul><h2>Documentation</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Architecture documentation added</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Retraining workflow documented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model versioning guide documented</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> MLOps operational guide added</p></li></ul><hr><h1>📊 Issue Metadata</h1><p><strong>Difficulty:</strong> Advanced</p><p><strong>Labels:</strong> <code inline="">ai</code> <code inline="">ml</code> <code inline="">mlops</code> <code inline="">backend</code> <code inline="">automation</code></p><p><strong>Impact Score:</strong> 9/10</p><p><strong>Effort Score:</strong> 8/10</p><p><strong>Priority:</strong> 🔴 HIGH</p><hr><h2>Additional Notes</h2><p>This feature transforms HelpDesk.AI from a static machine-learning application into a continuously improving AI platform. By leveraging administrator feedback, automated retraining, active learning strategies, and safe deployment practices, the classifier can evolve alongside customer needs while maintaining operational reliability and enterprise-grade governance.</p></body></html>
</body>
</html>

**Comment by riteshbonthalakoti** [comment]
Hi GSSoC contributors! 🙌

Thank you so much for your interest in contributing to HELPDESK.AI! This issue is open and available for ALL contributors! 🚀

Under GSSoC rules, multiple contributors can work on the same issue and the best, most comprehensive work will get merged. 

If you want to tackle this, please complete the mandatory onboarding steps below and start coding! 🚀💻


### 🌟 Mandatory Onboarding Steps (Must Complete!)
To ensure your GSSoC contribution is evaluated and your account cleared for dashboard testing, please complete these steps:
1. ⭐ **Star this repository**: https://github.com/ritesh-1918/HELPDESK.AI
2. 🍴 **Fork this repository**: https://github.com/ritesh-1918/HELPDESK.AI/fork
3. 👤 **Follow @ritesh-1918 on GitHub**: https://github.com/ritesh-1918
4. 💼 **Connect on LinkedIn**: https://www.linkedin.com/in/ritesh1908/
5. 🚀 **Register & Onboard**:
   - Go to the deployed app: https://helpdeskaiv1.vercel.app/
   - Click **Sign In** -> **Create Account**
   - Select **Ritesh Private Limited Company** as your organization.
   - Reply to this thread or email ritesh at bonthalamadhavi1@gmail.com with your username to get your access approved instantly!

*Note: All PR branches must target the `gssoc` branch, NOT `main`.*

**Comment by riteshbonthalakoti** [comment]
Resolved by merging all relevant GSSoC contributions. Closing issue. Thank you!
